Applications/FileIO/SWMM/SWMMInterface.cpp

```diff
@@ -0,0 +1,1248 @@
+: _base_name (swmm_base_name), _mesh(nullptr)
+{
+}
+
+SwmmInterface::~SwmmInterface()
+{
+    for (Subcatchment sc  : _subcatchments)
+        delete sc.outline;
+
+    for (GeoLib::Point* pnt : _subcatchment_points)
+        delete pnt;
+}
+
+bool SwmmInterface::isSwmmInputFile(std::string const& inp_file_name)
+{
+    std::string path (inp_file_name);
+    std::transform(path.begin(), path.end(), path.begin(), tolower);
+    if (BaseLib::getFileExtension(path) != "inp")
+    {
+        ERR ("SWMMInterface: file extension %s not recognised (should be *.inp).",
+            BaseLib::getFileExtension(inp_file_name).c_str());
+        return false;
+    }
+
+    std::ifstream in ( inp_file_name.c_str() );
+    if (!in.is_open())
+    {
+        ERR ("SWMMInterface: Could not open input file %s.", inp_file_name.c_str());
+        return false;
+    }
+
+    std::string line;
+    bool header_found (false);
+    std::size_t pos_beg (0), pos_end (0);
+    while (!header_found)
+    {
+        getline(in, line);
+        pos_beg = line.find_first_not_of(' ', pos_end);
+        pos_end = line.find_first_of(" \n", pos_beg);
+
+        if (line.empty() || pos_beg==pos_end || line.compare(pos_beg,2,";;") == 0)
+            continue;
+
+        if (line == "[TITLE]")
+            header_found = true;
+        else
+        {
+            ERR ("SWMMInterface: input file type %s not recognised.",
+                BaseLib::getFileExtension(inp_file_name).c_str());
+            return false;
+        }
+    }
+
+    in.close();
+    return true;
+}
+
+bool SwmmInterface::isSwmmOutputFile(std::string const& out_file_name)
+{
+    std::string path (out_file_name);
+    std::transform(path.begin(), path.end(), path.begin(), tolower);
+    if (BaseLib::getFileExtension(path) != "out")
+    {
+        ERR ("SWMMInterface: file extension %s not recognised (should be *.out).",
+            BaseLib::getFileExtension(out_file_name).c_str());
+        return false;
+    }
+
+    std::ifstream in ( out_file_name.c_str() );
+    if (!in.is_open())
+    {
+        ERR ("SWMMInterface: Could not open input file %s.", out_file_name.c_str());
+        return false;
+    }
+    in.close();
+    return true;
+}
+
+bool SwmmInterface::readPolygons(std::ifstream &in, std::vector<GeoLib::Polyline*> &lines,
+    std::vector<std::string> &ply_names, std::vector<GeoLib::Point*> &points,
+    std::vector<std::string> &pnt_names)
+{
+    bool finished (false);
+    std::size_t id (points.size());
+    std::string line;
+    std::string polygon_name("");
+    GeoLib::Polyline* p (nullptr);
+    getline(in, line);
+    while (!isSectionFinished(line))
+    {
+        if (isCommentLine(line))
+        {
+            getline(in, line);
+            continue;
+        }
+
+        std::vector<std::string> split_str (BaseLib::splitString(line));
+        if (split_str.size() != 3)
+        {
+            ERR ("Polygon format not recognised.");
+            return false;
```

Comment:
I think that doesn't work. I add p to the lines vector. When the method is finished, unique pointers will be cleaned and the polygon in the vector will change to nullptr or undefined or something. Also, at this point the worst that could happen is that p is a nullptr. Otherwise it's already been added to the vector.